source/common/router/upstream_request.cc

```diff
@@ -515,6 +528,7 @@ void UpstreamRequest::onPoolReady(
         [this]() -> void { onStreamMaxDurationReached(); });
     max_stream_duration_timer_->enableTimer(*max_stream_duration);
   }
+  copyDynamicMetaDataFromDownstream(info);
```

Comment:
@mattklein123, let me explain more for my cases and solution.

now client1 and client2 send HTTP requests to listener1 and share the same upstream connection to forward the HTTP requests to upstream servers. 

I want to use tcpdump to only capture the client1's traffic including  both downstream and upstream traffic. I do NOT want to capture any  downstream/upstream traffic for client2.

Currently I only know client1's IP address and listener1, I even do not know the upstream server's IP/port. how can I capture the traffic I want?
It is supposed that there are several downstream connections (HTTP protocol)for the listener.

downstream connection1:    client1(1.1.1.1:port1)<------->listener1(2.2.2.2:80)
downstream connection2:    client1(1.1.1.1:port2)<------->listener1(2.2.2.2:80)
downstream connection3:    client2(1.1.1.2:port3)<------->listener1(2.2.2.2:80)

There is one upstream connection (HTTP protocol) associate the above downstream connections.

envoy(3.3.3.3:port4)<------> upstream server (4.4.4.4:8080)

With tcpdump, it is easy to capture downstream traffic between client1 and listener1. But it is impossible to capture the upstream traffic as needed.

1. We do not know the upstream server IP since there are many upstream servers and also there may be L7 route policy to forward the traffic to many clusters.
2. Even if we know the specify upstream server's IP/port, we can not distinguish the traffic since client2 also use the same upstream connection.
3. Even if we can capture the upstream traffic only for client1, we can not distinguish the upstream traffic is for downstream connection1 or downstream connection2.

Now step back to the solution. I believe that when envoy send a packet(HTTP request) to upstream servers, envoy should know the packet is triggered by which downstream connection, right?
This PR is just solve this issue to let upstream connection know the associated downstream connection. 

First, the packet trace filter will set dynamical metadata for client1 traffic instead of client2 traffic, the client port info (port1/port2) will be saved in dynamic metadata.

Then, in UpstreamRequest::onPoolReady, copy the dynamic metadata to upstream streamInfo. If it is for downstream connection1, the dynamic metadata will be set to port1, if it is for downstream connection2, it will be set to port2. This dynamic metadata will change continuously for downstream connection1 and connection2.

Last, when envoy call IoResult RawBufferSocket::doWrite() to send packet,  I will use sendmsg() instead of write().  The sendmsg code is like
```
  const size_t cmsg_space = CMSG_SPACE(sizeof(uint32_t));
  absl::FixedArray<char> cbuf(cmsg_space);
  memset(cbuf.begin(), 0, cmsg_space);
  message.msg_control = cbuf.begin();
  message.msg_controllen = cmsg_space;
  cmsghdr* cmsg = CMSG_FIRSTHDR(&message);
  cmsg->cmsg_level = SOL_SOCKET;
  cmsg->cmsg_type = SO_MARK;
  cmsg->cmsg_len = CMSG_LEN(sizeof(uint32_t));
  memcpy(CMSG_DATA(cmsg), &mark, sizeof(mark));  // **mark is port1 or port2**
  const Api::SysCallSizeResult result = os_syscalls.sendmsg(fd_, &message, 0);
  ```
  
Now the port1/port2 is passed to kernel side. 
In kernel side, I will add a net filter module to add the port1/port2 as the tcp options in the payload.  Then tcpdump can recognize it.

Reply:
So you are planning on using a custom transport socket to achieve this?

The problem is that your solution will only work for HTTP/1. It won't work for any multiplexing protocol. As such I don't think we would accept a patch like this upstream.

I think your use case is probably better served by private extensions. I would focus more on what if any upstream Envoy changes are required to support your extensions - though even then it's not immediately obvious to me how to make this acceptable in a generic way that would not be specific to a non-multiplexing protocol.

Reply:
"So you are planning on using a custom transport socket to achieve this?" 
-- Yes. Need another PR.

"The problem is that your solution will only work for HTTP/1. It won't work for any multiplexing protocol"
-- For multiplexing protocol such as HTTP2, multiple downstream requests may be merged into one request for upstream. Is it possible to add all downstream requests's port info as an array in dynamic metadata ?

Reply:
I think it might be something you could make work with adding to a list, but at a certain point these seems very specific to your use case and I would just keep your extensions private and not upstream them. I would focus only on the portion required to move downstream state to upstream, and see if we can find a design that we would accept upstream. So I would just remove all of the filter and transport socket stuff and focus on that.

Reply:
"keep your extensions private and not upstream them", what does that mean ? Create a new branch for that ?

Reply:
I mean to keep them private to your internal build. It's not clear to me this is a feature that we would publicly accept as it seems HTTP/1 specific. 

Another way of thinking about this that we might accept would be to somehow build this into the existing tap filter infrastructure.

Reply:
I do not think it is HTTP/1 specific. It can also work for HTTP2 if we add the multiple downstrean info into a list. And also it works for pure TCP. 

For tap filter,  in my opinion, it is used to mirror traffic to another place. It seems that it is not for my case.

Reply:
I think at this point we need to move to a design doc. Can you please write a gdoc with a problem overview, solutions considered, proposed solution, etc.? I think that will be more productive at this point. Thank you.